ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,351 @@
+      Xapi_slave_db.clear_db ()
+    end;
+  if not (created_time_stamp_sanity !Xapi_slave_db.slave_db tables && modified_time_stamp_sanity !Xapi_slave_db.slave_db tables) then begin
+    debug "The stat fields are not correct - clearing slave db";
+    Xapi_slave_db.clear_db ()
+  end;
+  tables
+
+let send_notification __context token (u:update) =
+  let __context = Xapi_slave_db.update_context_db __context in
+  let reasons = ref [] in
+  if token < u.stat.created && u.stat.created < u.stat.modified then
+    reasons := ["add"; "mod"];
+  if token < u.stat.created && u.stat.created = u.stat.modified then
+    reasons := ["add"];
+  if u.stat.created < token && token < u.stat.modified then
+    reasons := ["mod"];
+  if u.stat.modified = u.stat.deleted then
+    reasons := [];
+
+  if u.fldname = "" then List.iter (fun r -> debug "Reason: %s" r) !reasons;
+  List.iter (fun r -> Db_action_helper.events_notify u.tblname r u.objref) !reasons
+
+let handle_notifications __context token (master_delta:delta) tables =
+  let dels = ref [] in
+  Mutex.execute Xapi_slave_db.slave_db_mutex (fun () ->
+      dels := get_deleted token !Xapi_slave_db.slave_db tables
+    );
+  List.iter (fun (d:del_tables) -> Db_action_helper.events_notify d.table "del" d.key) !dels;
+  List.iter (send_notification __context token) (f_all master_delta.tables)
+
+let write_db_to_disk =
+  if !Xapi_globs.slave_dbs then
+    begin
+      debug "Writing slave db to slave disk";
+      Db_cache_impl.sync (Db_conn_store.read_db_connections ()) !Xapi_slave_db.slave_db
+    end
+
+let start_stunnel_connection () =
+  Master_connection.is_slave := (fun () -> Pool_role.is_slave ());
+  Master_connection.get_master_address := Pool_role.get_master_address;
+  Master_connection.Slave_backup_connection.master_rpc_path := Constants.database_backup_uri;
+  Master_connection.Slave_backup_connection.on_database_connection_established :=
+    (fun () -> Master_connection.get_master_address := Pool_role.get_master_address);
+  Master_connection.Slave_backup_connection.start_master_connection_watchdog ();
+  Master_connection.Slave_backup_connection.open_secure_connection ();
+  debug "Created connection to master"
+
+let loop ~__context () =
+  let delay_seconds = !Xapi_globs.slave_db_writeout_time in
+  let start = Mtime_clock.counter () in
+  let token = ref "-2" in
+  let psec = Xapi_globs.pool_secret in
+  let addr = ref (Pool_role.get_master_address()) in
+  while (true) do
+    let new_master_address = Pool_role.get_master_address () in
+    let new_pool_secret = !Xapi_globs.pool_secret in
+    if new_master_address <> !addr || new_pool_secret <> !psec then begin
+      debug "The address or pool secret has changed - Make a new connection";
+      addr := new_master_address;
+      psec := new_pool_secret;
+      start_stunnel_connection ();
+    end;
+    let now = Mtime_clock.count start in
+    let changes =
```

Comment:
I don't think that this is sleeping on changes anymore. This will be a relatively tight loop querying the master

Reply:
This is fixed now